lib/auth/auth_with_roles.go

```diff
@@ -778,6 +781,81 @@ func (a *AuthWithRoles) DeleteWebSession(user string, sid string) error {
 	return a.authServer.DeleteWebSession(user, sid)
 }
 
+func (a *AuthWithRoles) GetAccessRequests(filter services.AccessRequestFilter) ([]services.AccessRequest, error) {
+	// An exception is made to allow users to get their own access requests.
```

Comment:
This was a little confusing for me to read, I was concerned I would miss some case in which the access control checks were not performed. What do you think about re-factoring to the following:

```go
// Allow users to get their own access requests, not access checks need to be performed.
if a.currentUserAction(filter.User) == nil {
   return a.authServer.GetAccessRequests(filter)
}

if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbList); err != nil {
	return nil, trace.Wrap(err)
}
if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbRead); err != nil {
	return nil, trace.Wrap(err)
}
return a.authServer.GetAccessRequests(filter)
```

Reply:
Also, does filter contain any other fields that need to be validated other than `filter.User`? For example what happens if I set the user to myself but set `filter.ID` to a request not owned by me?

Reply:
The re-factor proposed won't work because the `action` permissions need to only be checked if the `currentUserAction` check *fails*.  I'm definitely on-board with changing it, but I'm not 100% certain how to make it more clear.  The way its written now is consistent with how `currentUserAction` is used in the `GetUser` method, which also makes an exception where users can always access their own data.

> what happens if I set the user to myself but set filter.ID to a request not owned by me?

All filter fields must match, so this will return no requests since the only request with a matching `ID` does not have a matching `User`.

Reply:
The re-factor proposed won't work because the `action` permissions need to only be checked if the `currentUserAction` check *fails*.  I'm definitely on-board with changing it, but I'm not 100% certain how to make it more clear.  The way its written now is consistent with how `currentUserAction` is used in the `GetUser` method, which also makes an exception where users can always access their own data.

> what happens if I set the user to myself but set filter.ID to a request not owned by me?

All filter fields must match, so this will return no requests since the only request with a matching `ID` does not have a matching `User`.